crates/wasi-common/src/sys/windows/hostcalls_impl/misc.rs

```diff
@@ -77,12 +77,158 @@ pub(crate) fn clock_time_get(clock_id: wasi::__wasi_clockid_t) -> Result<wasi::_
     duration.as_nanos().try_into().map_err(Into::into)
 }
 
+fn stdin_nonempty() -> bool {
+    use std::io::Read;
+    std::io::stdin().bytes().peekable().peek().is_some()
+}
+
 pub(crate) fn poll_oneoff(
     timeout: Option<ClockEventData>,
     fd_events: Vec<FdEventData>,
     events: &mut Vec<wasi::__wasi_event_t>,
-) -> Result<Vec<wasi::__wasi_event_t>> {
-    unimplemented!("poll_oneoff")
+) -> Result<()> {
+    use crate::fdentry::Descriptor;
+    if fd_events.is_empty() && timeout.is_none() {
+        return Ok(());
+    }
+
+    // Currently WASI file support is only (a) regular files (b) directories (c) symlinks on Windows,
+    // which are always ready to write on Unix.
+    //
+    // We need to consider stdin/stdout/stderr separately.
+    // We treat stdout/stderr as always ready to write. I'm not sure if it's correct
+    // on Windows but I have not find any way of checking if a write to stdout would block.
+    // Therefore, we only poll the stdin.
+    let mut stdin_events = vec![];
```

Comment:
Given the way we separate the requested events into two lists (one that contains requests to read stdin and all others), it's possible to effectively *never* "poll" `stdin` provided that there's at least one other type of requested event always passed in along with the request to read from stdin.

I think perhaps we should peek at stdin if requested to read it and *only* attempt the busy wait loop if it's not ready.

Perhaps an approach that loops `fd_events` once and services every request it can, including a read of `stdin` that is ready,  pushing to `events` as it goes? If there's a request to read `stdin` and `stdin` is not ready, then we can add the request to a separate list.  After that, if the "not ready stdin read" list is empty, we're done.  If not, we do the busy wait with timeout for stdin to be ready?

Reply:
You're perfectly right. On Linux `poll` will always return `stdin` if it's immediately ready for reading, even if there are other descriptors immediately ready. I'm going to fix it.